Eigensolver Performance Improvement

fdkong opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

1. Use the initial guess from the previous simulation. Closes #12767 

2. Add colorful outputs support. Print out the eigenvalue at each power iteration and the end of Newton solver.  Closes #14292

3.  Separate power iteration and Newton. Newton does not start from Power during Picard iteration.

Closes #15513

fdkong committed:
Add short cuts for eigenvalue solvers

Refs #15513
Closes #15739

fdkong committed:
Eigen solver: set an initial guess to Newton

Closes #15513

fdkong committed:
Attempted to sperate nonlinear solver into two parts: nonlinear power and Newton.

Closes #15513

fdkong committed:
Newton correctly takes an initial guess from free power iterations

Closes #15513

fdkong committed:
Eigenvalue output and colorful residuals

Closes #14292

fdkong commented:
Require libmeh update https://github.com/libMesh/libmesh/pull/2697

fdkong commented:
@YaqiWang Here is the draft.

With this PR, 

1) we are able to reuse initial guess from the previous simulation 

2) we have colorful outputs. 

We need to add docs and more regression tests.

fdkong committed:
Add an option to output the eigenvalue as its inverse.

It is useful for neutron calculations. The inverse of the eigenvalue is the multiplication factor.

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -51,6 +51,11 @@ Eigenvalue::validParams()
   params.addParam<Real>(
       "normal_factor", 1.0, "Normalize eigenvector to make a defined norm equal to this factor");
 
+  params.addParam<bool>("auto_initialization",
```
This parameter is not like the one in the old eigenvalue executioner. The old eigenvalue executioner does not have slepc. The default value of this parameter is true. When this parameter is set to true, MOOSE will initialize the eigenvector to const 1. And this initialization is done in Executioner init. When `execute` is called again, initialization will not be done again.

fdkong replied:
We do not want to do `initialization` in the second `execute` because you want to reuse the solution from the previous step in Picard iteration.

YaqiWang commented on framework/src/systems/NonlinearEigenSystem.C:
```diff
@@ -239,6 +239,13 @@ NonlinearEigenSystem::solve()
+  if (_eigen_problem.needInitializeEigenVector())
+  {
+    _eigen_problem.initEigenvector(1.0);
+    _transient_sys.set_initial_space(solution());
```
If we want to have the behavior in the old executioner, we should call this line always. and move the above line in `Eigenvalue::init`.

fdkong replied:
It is always called. `if` is to check that it is indeed a nonlinear solver. We care about nonlinear solver at this point.

YaqiWang commented on framework/src/problems/EigenProblem.C:
```diff
@@ -316,6 +316,26 @@ EigenProblem::scaleEigenvector(const Real scaling_factor)
+        std::set<dof_id_type> var_indices;
+        _nl_eigen->system().local_dof_indices(var.number() + vc, var_indices);
+        for (const auto & dof : var_indices)
+          _nl_eigen->solution().set(dof, initial_value);
```
I know this code is the same as the one in the old executioner, but I'd let you be aware that this is not very right. Setting all dofs to this initial value does not make the solution flat in cases of MONOMIAL. It is right for LAGRANGE, but not always right. I was sloppy during the old implementation because libMesh does not have a simple way for us to set dofs so that the solution is constant over the domain.

fdkong replied:
I will check `MONOMIAL`

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -97,10 +97,23 @@ Eigenvalue::init()
+  Steady::execute();
+}
+
+void Eigenvalue::prepareSolverOptions()
```
No one is calling this function?

fdkong replied:
See line 117.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -106,6 +106,8 @@ getSlepcEigenProblemValidParams()
 
   params.addParam<unsigned int>("free_power_iterations", 4, "The number of free power iterations");
 
+  params.addParam<unsigned int>("extra_power_iterations", 0, "The number of extra free power iterations");
```
Where this parameter is used?

fdkong replied:
In Eigenvalue::excute

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -52,13 +52,12 @@ Eigenvalue::validParams()
-  params.addParam<bool>("newton_inverse_power",
-                        true,
-                        "If Newton and Inverse Power is combined in SLEPc side");
+  params.addParam<bool>(
```
Why is this parameter?  I thought for the slepc solver inside Eigenvalue::execute, we always have the combined newton and inverse power. On the other hand, the solver in Eigenvalue::init, we only want to have fixed number of inverse power iterations.

fdkong replied:
This parameter is added when we want SLEPc to do everything instead of separating `power` and `newton`.

I believe  Grifin will not touch this parameter

fdkong replied:
addprivateparam

YaqiWang commented on framework/include/problems/EigenProblem.h:
```diff
@@ -143,6 +143,10 @@ class EigenProblem : public FEProblemBase
 
+  bool outputInverseEigenvalue() { return _output_inverse_eigenvalue; }
+
+  void outputInverseEigenvalue(bool inverse) { _output_inverse_eigenvalue = inverse; }
```
did we decide to have a static function that GriffinApp can call to change the behavior of printing eigenvalues at the end?

fdkong replied:
I knew that. But I realized it was not necessary to be static. This function does not need to be set earlier. 

But I could do that

YaqiWang replied:
Then how Griffin change the default behavior? We do not want to ask users to set that parameter in the input all the time though.

fdkong replied:
Suppress parameter or call this method in SNProblem.

YaqiWang reviewed (changes requested):
@fdkong seems you have some misunderstandings.

fdkong commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -89,10 +110,71 @@ Eigenvalue::init()
 Eigenvalue::execute()
+{
+  // Let us do extra power iterations here if necessary
+  auto extra_power_iterations = _pars.get<unsigned int>("extra_power_iterations");
```
getParams

YaqiWang commented:
Just had a discussion with Fande on phone. I was looking at some outdated codes... We looked the code together and Fande will tag me once our discussions are addressed.

fdkong committed:
Eigen solver: updated picard example

Closes #15513

fdkong committed:
Eigenvalue derives from Executioner instead of Steady

fdkong committed:
Add a test using "extra_power_iterations" option

fdkong committed:
Autodetect eigen variables.

If a variable operates on Eigen kernels,  we mark the variable as an Eigen variable

pbalest commented:
Quick question how this is going to work? it will be a flag to turn on or the eigenvalue will use the old solution as initial guess every iteration by default?

fdkong commented:
> Quick question how this is going to work? it will be a flag to turn on or the eigenvalue will use the old solution as initial guess every iteration by default? can I start to try it or it need further work?

If `eigenvalue` is the master, it will take the previous solution by default. A good example is `/test/tests/problems/eigen_problem/eigensolvers/ne_coupled_picard.i`. 

If `eigenvalue` is the subapp, you need ` keep_solution_during_restore = true` in `MultiApps` block to have the subapp use the previous solution as the initial guess. The good example is `moose/test/tests/problems/eigen_problem/eigensolvers/ne_coupled_picard_subT.i`

pbalest commented:
Do I need this PR in? Because I think I already tried with `keep_solution_during_restore = true`, let me double check!

fdkong commented:
> Do I need this PR in? Because I think I already tried with `keep_solution_during_restore = true`, let me double check!

Yes, you need this PR.

pbalest commented:
Nice! I'll give a try then! thanks @fdkong

pbalest commented:
@fdkong is this branch compatible with the latest griffin version? Do I need a particular Libmesh version?
```
Compiling C++ (in opt mode) /Users/balep/projects/griffin-idaholab-new1/moose/framework/build/unity_src/outputs_formatters_Unity.C...
In file included from /Users/balep/projects/griffin-idaholab-new1/moose/framework/build/unity_src/systems_Unity.C:8:
/Users/balep/projects/griffin-idaholab-new1/moose/framework/src/systems/NonlinearEigenSystem.C:282:20: error: no member named 'set_initial_space' in 'libMesh::TransientSystem<libMesh::EigenSystem>'
    _transient_sys.set_initial_space(solution());
```

fdkong commented:
> @fdkong is this branch compatible with the latest griffin version? Do I need a particular Libmesh version?
> 
> ```
> Compiling C++ (in opt mode) /Users/balep/projects/griffin-idaholab-new1/moose/framework/build/unity_src/outputs_formatters_Unity.C...
> In file included from /Users/balep/projects/griffin-idaholab-new1/moose/framework/build/unity_src/systems_Unity.C:8:
> /Users/balep/projects/griffin-idaholab-new1/moose/framework/src/systems/NonlinearEigenSystem.C:282:20: error: no member named 'set_initial_space' in 'libMesh::TransientSystem<libMesh::EigenSystem>'
>     _transient_sys.set_initial_space(solution());
> ```

We need

https://github.com/libMesh/libmesh/pull/2697

fdkong commented:
You could check my submodule commit.

pbalest commented:
Just saw it thanks! I was using the conda libmesh that's probably why got that error!

pbalest commented:
I am trying an input file that was running before on HPC, with this version do not run anymore.. this is what I get:
```
*** ERROR ***
It is not eps power, and there is no snes

[cli_0]: aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

fdkong commented:
> I am trying an input file that was running before on HPC, with this version do not run anymore.. this is what I get:
> 
> ```
> *** ERROR ***
> It is not eps power, and there is no snes
> 
> [cli_0]: aborting job:
> application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
> ```
> 
> These are the loaded modules:
> 
> ```
> use.moose   2) CMake/3.9.4   3) python/3.7-anaconda   4) GCC/8.2.0   5) MVAPICH2/2.3.1-GCC   6) PETSc/3.11.4-GCC
> ```

What the executioner block looks like?

It is better to use petsc submodule (3.13). I did not check the code against the old PETSc yet.

pbalest commented:
```
[Executioner]
  type = Eigenvalue
  matrix_free = true
  solve_type = Newton
  eigen_problem_type = GEN_NON_HERMITIAN
  free_power_iterations = 4
  nl_abs_tol = 1e-9
  auto_preconditioning = false
[]
```

fdkong commented:
> [Executioner]
>   type = Eigenvalue
>   matrix_free = true
>   solve_type = Newton
>   eigen_problem_type = GEN_NON_HERMITIAN
>   free_power_iterations = 4
>   nl_abs_tol = 1e-9
>   auto_preconditioning = false
> []

Could you try petsc submodule?

pbalest commented:
Trying but it will be easy because HPC is can't download the missing libraries... looking for an option to transfer them manually

fdkong commented:
> Trying but it will not be easy because HPC is can't download the missing libraries... looking for an option to transfer them manually

Offline installation instructions at https://github.com/idaholab/moose/pull/15765 may help. It will instruct you to install moose on a machine that does not have internet access.

pbalest commented:
Thanks  I was trying to access the website but, is still down.. I'll compile the documentation and follow istructions!

YaqiWang committed:
refactor normalization for Eigenvalue to make it as part of problem solve #15807

fdkong committed:
Eigen solver: Remove extra option "-eps_power" from SNES

since users do not like it

fdkong committed:
Eigen solver: Remove extra option "-eps_power" from SNES

since users do not like it

fdkong committed:
Eigenvalue excutioner: nondimensionalize initial eigenvector

fdkong committed:
Eigenvalue excutioner: set ksp default norm type and pc side

fdkong committed:
Eigenvalue solver: scale solution back if it has been scaled in the previous solve

This will make solver run faster since the initial guess is better

fdkong committed:
Apply clang-format after rebase on origin/devel

fdkong commented:
@pbalest I start to working on the integration

fdkong commented:
libMesh update is needed. @lindsayad Where is the latest libmesh update?

lindsayad commented:
https://github.com/idaholab/moose/pull/15867

lindsayad commented:
I think @loganharbour had a problem where draft-prs don't update conflicts?

fdkong commented:
> #15867

` _transient_sys.set_initial_space(solution());` should be in, and I am puzzled why civet did not see the update

YaqiWang commented on framework/include/problems/EigenProblem.h:
```diff
@@ -105,12 +116,77 @@ class EigenProblem : public FEProblemBase
+   * We need to mark the solver as "converged" when doing free power to retrieve
+   * the final solution from SLPEc
+   */
+  bool doFreePowerIteration() { return _do_free_power_iteration; }
```
`const`?

YaqiWang commented on framework/include/problems/EigenProblem.h:
```diff
@@ -105,12 +116,77 @@ class EigenProblem : public FEProblemBase
+  /**
+   * Return console handle, and we use that in EPSMonitor to print out eigenvalue
+   */
+  const ConsoleStream & console() { return _console; }
```
`const`?

lindsayad replied:
:+1: to that

fdkong replied:
Yes

YaqiWang commented on framework/include/utils/MooseTypes.h:
```diff
@@ -677,7 +677,9 @@ enum EigenSolveType
   EST_JACOBI_DAVIDSON, ///< Jacobi-Davidson
   EST_NONLINEAR_POWER, ///< Nonlinear inverse power
-  EST_NEWTON,          ///< Newton-based eigen solver
+  EST_NEWTON, ///< Newton-based eigensolver with an assembled Jacobian matrix (fully coupled by default)
```
What do you mean `fully coupled by default`? You mean the matrix coupling pattern is full?

fdkong replied:
Yes

fdkong replied:
You might never use it in griffin

YaqiWang replied:
hehe, I am actually going to use it.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/592269/) on 623f260 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15791/site)

This comment will be updated on new commits.

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -61,40 +77,228 @@ Eigenvalue::validParams()
+    // and users does not provide an initial guess, slepc will automatically generate
+    // a random vector as the initial guess. The motivation to offer this option is
+    // that we have to initialize  ONLY eigen variables in multiphysics simulation.
+    if (_eigen_problem.needInitializeEigenVector())
```
Should this be outside the `if` on line 162? User may set zero free power iterations but still want to do the initialization.

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -61,40 +77,228 @@ Eigenvalue::validParams()
+    _console << " Free power iteration starts" << std::endl;
+
+    // Call solver
+    _eigen_problem.solve();
```
I am getting `Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 1` in one of my test runs here. I suggest we clear the snes tolerances so that we do not get this.

fdkong replied:
Good suggestion

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -61,40 +77,228 @@ Eigenvalue::validParams()
+
+#ifdef LIBMESH_HAVE_SLEPC
+
+  // Make sure all PETSc options are setup correctly
```
Can we move the code in `init` from here into `_eigen_problem.initialSetup()`?

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -61,40 +77,228 @@ Eigenvalue::validParams()
+    for (MooseIndex(_num_grid_steps) grid_step = 0; grid_step <= _num_grid_steps; ++grid_step)
+    {
+
+      // Let us do extra power iterations here if necessary
```
From here to LIne 238, I think the code should be in `_eigen_problem.solve()`. We want the `PicardSolve` as the outermost iteration.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -920,6 +959,148 @@ PCSetUp_MoosePC(PC pc)
+  auto eigenvalue = inverse ? 1.0 / (*eigr) : (*eigr);
+
+  // The term "k-eigenvalue" is adopted from the neutronics community.
+  console << " Iteration " << its << std::setprecision(8)
```
Suggest to have precision 10 here. 8 is slightly on the margin i.e. nl_rel_tol may still not reached but eigenvalue is accurate in 8 precision.

fdkong replied:
OK

YaqiWang commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -24,7 +24,8 @@ Eigenvalue::validParams()
+      "Eigenvalue solves a standard/generalized linear or nonlinear eigenvaue problem");
 
   params.addParam<bool>(
       "matrix_free",
```
Do we still need this parameter here?

YaqiWang replied:
Now I am thinking based on our discussion that we should remove `precond_matrix_includes_eigen`. Users use kernels to decide how to assemble `A` and `B`. The code should manipulate `A` and `B` with `x` and `lambda` to form the full preconditioning matrix. We possibly can detect if `B` is zero or not. If `B` is zero, the manipulation is not needed.

fdkong replied:
matrix_free will be an invalid for griffin once the integration is done. In this PR, we can not change it.  It will still be a valid option when users  use non-Newton algorithms

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -106,6 +110,9 @@ getSlepcEigenProblemValidParams()
 
   params.addParam<unsigned int>("free_power_iterations", 4, "The number of free power iterations");
 
+  params.addParam<unsigned int>(
```
We need a protection here: out PJFNK eigen solver only can find the smallest eigenvalue, so we should error out if users set the `which_eigen_pairs` to something like `LARGEST_MAGNITUDE`. In more general sense, we need to check if a specific eigen solver in SLEPc can deliver the desired eigen pairs.

fdkong replied:
In general, we should not protect this in MOOSE, instead, it should be handled by SLEPc/PETSc.  For example, right now, PJFNK only makes sense for SMALLEST_MAGNITUDE (that was chosen by default). SLEPc will be angry we choose wrong parameters. In addition, in the future, PJFNK can be enhanced  to find other spectrums.

YaqiWang replied:
Is SLEPc erroring out on the wrong settings then?

fdkong replied:
I think so, otherwise thing is really wrong

YaqiWang committed:
do not zero out jacobian when moose has constant Jacobian and skips jacobian evaluations

fdkong commented:
OK, I am finally back this PR again

fdkong committed:
Output eigenvalue in precision-10

fdkong committed:
MOOSE can compile when there is no SLEPc

fdkong committed:
Fixed non-unity build errors

moosebuild commented:
Job [Minimum gcc version](https://civet.inl.gov/job/611666/) on 04265d7 : invalidated by @fdkong

warning: &#39;dummyvar_for_registering_obj_BoxMarker13

fdkong committed:
Output initial conditions before involving SLEPc solver

fdkong committed:
Skip a few tests when using SLEPc older than 3.13.0

because solver can not take an initial guess from moose
and the solver may converge to different mode when including
B into the preconditioning matrix

fdkong committed:
mooseDeprecated old slepc versions and used libmesh_make_unique

lindsayad commented on framework/include/problems/EigenProblem.h:
```diff
@@ -57,7 +58,56 @@ class EigenProblem : public FEProblemBase
+  /*
+   * Specify whether or not to initialize eigenvector automatically
+   */
+  void needInitializeEigenVector(bool need) { _auto_initilize_eigen_vector = need; }
```
`initilize` -> `initialize`

fdkong replied:
Oh, my god!

lindsayad commented on framework/include/problems/EigenProblem.h:
```diff
@@ -57,7 +58,56 @@ class EigenProblem : public FEProblemBase
+   * Set postprocessor and normalization factor
+   * 'Postprocessor' is often used to compute an integral of physics variables
+   */
+  void setNormalization(const PostprocessorName pp,
```
`const PostprocessName &`

fdkong replied:
Yes

lindsayad commented on framework/include/problems/EigenProblem.h:
```diff
@@ -57,7 +58,56 @@ class EigenProblem : public FEProblemBase
+   * We need to mark the solver as "converged" when doing free power to retrieve
+   * the final solution from SLPEc
+   */
+  bool doFreePowerIteration() { return _do_free_power_iteration; }
```
Make method `const`

fdkong replied:
Yes

lindsayad commented on framework/include/problems/EigenProblem.h:
```diff
@@ -105,12 +155,37 @@ class EigenProblem : public FEProblemBase
   /**
    * Which eigenvalue is active
    */
   unsigned int activeEigenvalueIndex() { return _active_eigen_index; }
```
Not in the diff, but let's make this a `const` method

fdkong replied:
Yes

lindsayad commented on framework/include/problems/EigenProblem.h:
```diff
@@ -121,11 +196,23 @@ class EigenProblem : public FEProblemBase
 
   unsigned int _active_eigen_index;
+
+  bool _auto_initilize_eigen_vector;
```
`initilize` -> `initialize`

fdkong replied:
Yes

lindsayad commented on framework/include/problems/EigenProblem.h:
```diff
@@ -121,11 +196,23 @@ class EigenProblem : public FEProblemBase
   PerfID _compute_jacobian_blocks_timer;
+
+  bool _has_normalization;
+  PostprocessorName _normalization;
+  Real _normal_factor;
```
What is the difference between these two? It might be helpful to add some doxygen explaining what these individual members do

fdkong replied:
Yes

lindsayad commented on framework/include/systems/NonlinearEigenSystem.h:
```diff
@@ -92,6 +99,12 @@ class NonlinearEigenSystem : public NonlinearSystemBase
 
+  /**
+   * Retrieve snes from slepc eigen solver. It is valid for only nonlinear eigen solver.
+   * You should see a big error if you do this for linear solver.
```
Big error is better than small error :smile:

lindsayad commented on framework/include/systems/NonlinearEigenSystem.h:
```diff
@@ -125,7 +138,7 @@ class NonlinearEigenSystem : public NonlinearSystemBase
    * @return all converged eigenvalues as complex numbers
    */
-  virtual const std::vector<std::pair<Real, Real>> & getAllConvergedEigenvalues()
+  virtual const std::vector<std::pair<Real, Real>> & getAllConvergedEigenvalues() const
```
Oh I love `const`. Are you actually going to override this method somewhere? Wondering if this needs to be `virtual`. I always thought that `NonlinearEigenSystem` was a leaf object, but maybe you derive from it in Griffin or something.

lindsayad commented on framework/include/utils/SlepcEigenSolverConfiguration.h:
```diff
@@ -0,0 +1,46 @@
+  SlepcEigenSolverConfiguration(EigenProblem & eigen_problem,
+                                libMesh::SlepcEigenSolver<libMesh::Number> & slepc_eigen_solver);
+
+  /**
+   * empty destructor
+   */
+  ~SlepcEigenSolverConfiguration() {}
```
If it's empty, then we do not need

lindsayad commented on framework/include/utils/SlepcSupport.h:
```diff
@@ -53,6 +56,26 @@ PetscErrorCode mooseSlepcEigenFormJacobianB(SNES snes, Vec x, Mat jac, Mat pc, v
 PetscErrorCode mooseSlepcEigenFormFunctionA(SNES snes, Vec x, Vec r, void * ctx);
 PetscErrorCode mooseSlepcEigenFormFunctionB(SNES snes, Vec x, Vec r, void * ctx);
 PetscErrorCode mooseSlepcEigenFormFunctionAB(SNES snes, Vec x, Vec Ax, Vec Bx, void * ctx);
+PetscErrorCode mooseSlepcStoppingTest(EPS eps,
+                                      PetscInt its,
+                                      PetscInt max_it,
+                                      PetscInt nconv,
+                                      PetscInt nev,
+                                      EPSConvergedReason * reason,
+                                      void * ctx);
+PetscErrorCode mooseSlepcEPSGetSNES(EPS eps, SNES * snes);
+PetscErrorCode mooseSlepcEPSMonitor(EPS eps,
+                                    int its,
+                                    int nconv,
+                                    PetscScalar * eigr,
+                                    PetscScalar * eigi,
+                                    PetscReal * errest,
+                                    int nest,
+                                    void * mctx);
+
+PetscErrorCode mooseSlepcEPSSNESSetUpOptionPrefix(EPS eps);
+PetscErrorCode mooseSlepcEPSSNESSetCustomizePC(EPS eps);
+PetscErrorCode mooseSlepcEPSSNESKSPSetPCSide(FEProblemBase & problem, EPS eps);
```
It looks like have a history of not documenting these APIs, but I would love to see some doxygen here. Most of the method names are pretty self explanatory, but for instance I don't know just by looking at it what `mooseSlepcStoppingTest` does

fdkong replied:
Will do docs

lindsayad commented on framework/include/variables/MooseVariableBase.h:
```diff
@@ -144,6 +144,16 @@ class MooseVariableBase : public MooseObject,
    */
   virtual unsigned int numberOfDofs() const { return _dof_indices.size(); }
 
+  /**
+   * Whether or not this variable operates on an eigen kernel
+   */
+  virtual bool eigen() const { return _is_eigen; }
+
+  /**
+   * Mark this variable as an eigen var or non-eigen var
+   */
+  virtual void eigen(bool eigen) { _is_eigen = eigen; }
```
These are new methods with a base implementation that I don't see overridden anywhere. Can we remove `virtual`?

fdkong replied:
will do

lindsayad commented on framework/src/auxkernels/NormalizationAux.C:
```diff
@@ -37,6 +37,7 @@ Real
 NormalizationAux::computeValue()
 {
   Real denominator = _pp_on_source ? *_pp_on_source : 1.0;
+  mooseAssert(denominator != 0., "postprocessor value is zero");
```
Oh I love asserts

fdkong replied:
I will become another  Roy soon

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -61,40 +78,133 @@ Eigenvalue::validParams()
+
+  if (isParamValid("normalization"))
+  {
+    auto normpp = getParam<PostprocessorName>("normalization");
```
Not important but could be `const auto &` to avoid copies. I know that string is not very expensive to copy :smile:

fdkong replied:
Haha. There is a lesson for me, I should make everything constant. And then git rid of these that can not be made as constant

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -61,40 +78,133 @@ Eigenvalue::validParams()
+#endif
 }
 
+#if LIBMESH_HAVE_SLEPC
```
Not a huge deal but `#ifdef` convention. We either define it or we don't.

fdkong replied:
will do

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -114,31 +224,160 @@ Eigenvalue::execute()
+void
+Eigenvalue::execute()
+{
+#if LIBMESH_HAVE_SLEPC
```
`#ifdef`

fdkong replied:
will do

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -61,40 +78,133 @@ Eigenvalue::validParams()
+
+  if (isParamValid("normalization"))
+  {
+    auto normpp = getParam<PostprocessorName>("normalization");
```
`const auto &`. You possibly will never ask me to review again :rofl:

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -114,31 +224,160 @@ Eigenvalue::execute()
+void
+Eigenvalue::setFreeNonlinearPowerIterations(unsigned int free_power_iterations)
+{
+#if !PETSC_RELEASE_LESS_THAN(3, 12, 0)
```
This is a situation where `#if` is appropriate because we care about the value of the macro, not simply about whether it exists

fdkong replied:
It is really hard for me to get something right, haha

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -114,31 +224,160 @@ Eigenvalue::execute()
+#ifdef LIBMESH_ENABLE_AMR
+
+  // Define the refinement loop
+  unsigned int steps = _eigen_problem.adaptivity().getSteps();
```
Could use `auto` here in case you ever wanted to change the return type of `getSteps`. But I guess you will probably never do more than 2^32 steps :laughing:

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -114,31 +224,160 @@ Eigenvalue::execute()
+
+  // Define the refinement loop
+  unsigned int steps = _eigen_problem.adaptivity().getSteps();
+  for (unsigned int r_step = 0; r_step <= steps; r_step++)
```
You don't have to do this but I love using this relatively new libMesh utility function:
```c++
for (const auto r_step : make_range(steps))
```

fdkong replied:
Python style .

lindsayad commented on framework/src/executioners/Eigenvalue.C:
```diff
@@ -114,31 +224,160 @@ Eigenvalue::execute()
-      mooseError("Cannot normalize eigenvector by 0");
-    else
-      val /= *_normalization;
+    // This loop is for nonlinear multigrids (developed by Alex)
+    for (MooseIndex(_num_grid_steps) grid_step = 0; grid_step <= _num_grid_steps; ++grid_step)
```
I feel honored that you put my code here

fdkong replied:
haha

lindsayad commented on framework/src/problems/EigenProblem.C:
```diff
@@ -348,7 +357,123 @@ EigenProblem::scaleEigenvector(const Real scaling_factor)
+  std::vector<VariableName> var_names = getVariableNames();
+
+  // Count how many dofs we have
+  for (auto & vn : var_names)
+  {
+    auto & var = getVariable(0, vn);
```
You know that I love `const auto &`

lindsayad commented on framework/src/systems/NonlinearEigenSystem.C:
```diff
@@ -101,12 +102,25 @@ NonlinearEigenSystem::NonlinearEigenSystem(EigenProblem & eigen_problem, const s
   sys().attach_assemble_function(Moose::assemble_matrix);
 
+  SlepcEigenSolver<Number> * solver =
+      libmesh_cast_ptr<SlepcEigenSolver<Number> *>(&(*_transient_sys.eigen_solver));
```
No need to have back-to-back &*; they cancel each other. Unless `eigen_solver` is a smart ptr, then you can just call `.get()`

fdkong replied:
Will do

lindsayad commented on framework/src/systems/NonlinearEigenSystem.C:
```diff
@@ -155,6 +169,11 @@ NonlinearEigenSystem::initialSetup()
     addPrecondTagToMooseObjects(_integrated_bcs);
   }
+
+  // Mark a variable an eigen variable if it operates on eigen kernels
+  markEigenVariables(_dg_kernels);
+  markEigenVariables(_kernels);
+  markEigenVariables(_scalar_kernels);
```
This looks tricky. IIRC we cannot do eigen with (integrated) bcs? In which case, I assume it wouldn't make sense either to do them with interface kernels? I think this probably doesn't make sense thinking about the strong form of the eigenvalue problem

fdkong replied:
We support `integrated bcs` now.` interface kernels` should make sense, but there is no use case yet.

lindsayad commented on framework/include/systems/NonlinearEigenSystem.h:
```diff
@@ -76,6 +77,12 @@ class NonlinearEigenSystem : public NonlinearSystemBase
   template <typename T>
   void addPrecondTagToMooseObjects(MooseObjectTagWarehouse<T> & warehouse);
 
+  /**
+   * Mark a variable an eigen variable if it operates on eigen kernels
+   */
+  template <typename T>
+  void markEigenVariables(MooseObjectTagWarehouse<T> & warehouse);
```
If you are only going to use this in the `NonlinearEigenSystem` class, then you should make this method private. You put the definition in the .C file so if someone else attempted to call this we would end up with undefined references

fdkong replied:
Nice

lindsayad commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -122,29 +129,28 @@ setSlepcEigenSolverTolerances(EigenProblem & eigen_problem, const InputParameter
     // nonlinear solver tolerances
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_snes_max_it",
+    Moose::PetscSupport::setSinglePetscOption("-snes_max_it",
                                               stringify(params.get<unsigned int>("nl_max_its")));
 
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_snes_max_funcs",
+    Moose::PetscSupport::setSinglePetscOption("-snes_max_funcs",
                                               stringify(params.get<unsigned int>("nl_max_funcs")));
 
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_snes_atol",
+    Moose::PetscSupport::setSinglePetscOption("-snes_atol",
                                               stringify(params.get<Real>("nl_abs_tol")));
 
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_snes_rtol",
+    Moose::PetscSupport::setSinglePetscOption("-snes_rtol",
                                               stringify(params.get<Real>("nl_rel_tol")));
 
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_snes_stol",
+    Moose::PetscSupport::setSinglePetscOption("-snes_stol",
                                               stringify(params.get<Real>("nl_rel_step_tol")));
 
     // linear solver
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_ksp_max_it",
+    Moose::PetscSupport::setSinglePetscOption("-ksp_max_it",
                                               stringify(params.get<unsigned int>("l_max_its")));
 
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_ksp_rtol",
-                                              stringify(params.get<Real>("l_tol")));
+    Moose::PetscSupport::setSinglePetscOption("-ksp_rtol", stringify(params.get<Real>("l_tol")));
 
-    Moose::PetscSupport::setSinglePetscOption("-eps_power_ksp_atol",
+    Moose::PetscSupport::setSinglePetscOption("-ksp_atol",
```
This looks like a really nice change

fdkong replied:
haha. Get rid of prefix because nobody like it

lindsayad commented on modules/navier_stokes/test/tests/ins/IdealAMG/2d_diffusion_amg.i:
```diff
@@ -0,0 +1,48 @@
+[Mesh]
```
I'm guessing you didn't mean for this file to get in here?

lindsayad replied:
Beautiful cask branch?

fdkong replied:
It is disaster!!

lindsayad reviewed (changes requested):
This looks like an awesome PR @fdkong. My only requested changes are small C++ ones

fdkong commented on framework/include/outputs/OutputWarehouse.h:
```diff
@@ -194,6 +194,12 @@ class OutputWarehouse
   void reset();
 
+  /**
+   * Calls the timestepSetup function for each of the output objects
```
`timestepSetup` is too early for solver setup where I am trying to hook up monitor, change prefix, etc. SLEPc solver does not exist yet when we call `timestepSetup`.

I moved this method from the private region to the public region because I want to call this function from `EigenProblem`.

fdkong replied:
At the same time, I do not want to declare `EigenProblem` as friend of `OutputWarehouse`

lindsayad replied:
Adding some more doxygen for this reasoning would be appreciated, otherwise someone else like me might come along and be confused and think you made a typo and think that that method doesn't actually do what you say it does

fdkong replied:
Will do

fdkong commented on framework/include/problems/EigenProblem.h:
```diff
@@ -30,6 +30,7 @@ class EigenProblem : public FEProblemBase
 
   EigenProblem(const InputParameters & parameters);
 
+#if LIBMESH_HAVE_SLEPC
```
Will change. Thanks

fdkong commented on framework/include/problems/EigenProblem.h:
```diff
@@ -57,7 +58,56 @@ class EigenProblem : public FEProblemBase
+   * If we need to initialize eigen vector. We initialize the eigen vector
+   * only when "auto_initialization" is on and nonlinear eigen solver is selected.
+   */
+  bool needInitializeEigenVector();
```
Yes

fdkong committed:
Marked many functions as constant

Added more docs, and polished code

fdkong commented:
@lindsayad It is ready for review/merge

lindsayad reviewed (approved):
Oh my gosh, such beautiful documented and `const` code I have hardly ever seen

YaqiWang commented:
This is great! Thanks!
